runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkSplittableDoFnFunction.java

```diff
@@ -0,0 +1,274 @@
+      DoFn<KeyedWorkItem<K, V>, OutputT> dofn,
+      String stepName,
+      WindowingStrategy<?, ?> windowingStrategy,
+      Map<PCollectionView<?>, WindowingStrategy<?, ?>> sideInputs,
+      PipelineOptions pipelineOptions,
+      Map<TupleTag<?>, Integer> outputMap,
+      TupleTag<OutputT> mainOutputTag) {
+
+    this.dofn = dofn;
+    this.stepName = stepName;
+    this.windowingStrategy = windowingStrategy;
+    this.sideInputs = sideInputs;
+    this.serializedOptions = new SerializablePipelineOptions(pipelineOptions);
+    this.outputMap = outputMap;
+    this.mainOutputTag = mainOutputTag;
+  }
+
+  @Override
+  public void reduce(
+      Iterable<WindowedValue<KeyedWorkItem<K, V>>> values,
+      Collector<WindowedValue<OutputT>> out) throws Exception {
+    RuntimeContext runtimeContext = getRuntimeContext();
+
+    DoFnRunners.OutputManager outputManager;
+    if (outputMap.size() == 1) {
+      outputManager = new FlinkDoFnFunction.DoFnOutputManager(out);
+    } else {
+      // it has some additional Outputs
+      outputManager =
+          new FlinkDoFnFunction.MultiDoFnOutputManager((Collector) out, outputMap);
+    }
+
+    final Iterator<WindowedValue<KeyedWorkItem<K, V>>> iterator = values.iterator();
+
+    // get the first value, we need this for initializing the state internals with the key.
+    // we are guaranteed to have a first value, otherwise reduce() would not have been called.
+    WindowedValue<KeyedWorkItem<K, V>> currentValue = iterator.next();
+    final K key = currentValue.getValue().key();
+
+    final InMemoryStateInternals<K> stateInternals = InMemoryStateInternals.forKey(key);
+
+    // Used with Batch, we know that all the data is available for this key. We can't use the
+    // timer manager from the context because it doesn't exist. So we create one and advance
+    // time to the end after processing all elements.
+    final InMemoryTimerInternals timerInternals = new InMemoryTimerInternals();
+    timerInternals.advanceProcessingTime(Instant.now());
+    timerInternals.advanceSynchronizedProcessingTime(Instant.now());
+
+    FlinkSideInputReader sideInputReader = new FlinkSideInputReader(
+        sideInputs,
+        runtimeContext);
+
+    List<TupleTag<?>> additionalOutputTags = Lists.newArrayList(outputMap.keySet());
+
+    StateInternalsFactory<String> stateInternalsFactory =
+        k -> {
+          // this will implicitly be keyed by the key of the incoming
+          // element or by the key of a firing timer
+          return (StateInternals) stateInternals;
+        };
+    TimerInternalsFactory<String> timerInternalsFactory =
+        k -> {
+          // this will implicitly be keyed like the StateInternalsFactory
+          return timerInternals;
+        };
+
+    ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor(
+        Executors.defaultThreadFactory());
+
+    ((SplittableParDoViaKeyedWorkItems.ProcessFn) dofn)
```

Comment:
Extract into a variable?

Reply:
Yes! I realised we also don't have the generic type info. I'll try and fix that.